src/modules/wreck/job.c

```diff
@@ -151,97 +150,75 @@ static char * realtime_string (char *buf, size_t sz)
     return (buf);
 }
 
-static void wait_for_event (flux_t *h, int64_t id, char *topic)
-{
-    struct flux_match match = {
-        .typemask = FLUX_MSGTYPE_EVENT,
-        .matchtag = FLUX_MATCHTAG_NONE,
-    };
-    match.topic_glob = topic;
-    flux_msg_t *msg = flux_recv (h, match, 0);
-    flux_msg_destroy (msg);
-    return;
-}
-
-static void send_create_event (flux_t *h, int64_t id,
-                               const char *path, const char *state,
-                               json_object *req)
+/* Send wreck.state.<state> event.
+ * Instead of the usual "fire and forget" event interface, publish
+ * synchronously via the rank 0 cmb.pub service to ensure that response
+ * to job create request is not sent until the event has received a
+ * sequence number.  See issue #337.
+ */
+static flux_future_t *send_create_event (flux_t *h, struct wjob *job)
 {
-    int val;
-    int nnodes = 0;
-    int ntasks = 0;
-    int ncores = 0;
-    int walltime = 0;
-
-    char *topic;
-    flux_msg_t *msg;
-   if (asprintf (&topic, "wreck.state.%s", state) < 0) {
-        flux_log_error (h, "send_create_event: asprintf");
-        return;
-    }
+    char topic[64];
+    flux_future_t *f;
+    uint32_t nodeid = 0;
+    int flags = 0;
 
-    /* Pull ntasks, nnodes directly out of request
-     */
-    if (Jget_int (req, "ntasks", &val))
-        ntasks = val;
-    if (Jget_int (req, "nnodes", &val))
-        nnodes = val;
-    if (Jget_int (req, "ncores", &val))
-        ncores = val;
-    if (Jget_int (req, "walltime", &val))
-        walltime = val;
-
-    msg = flux_event_pack (topic, "{s:I,s:s,s:i,s:i,s:i,s:i}",
-                          "jobid", id, "kvs_path", path,
-                          "ntasks", ntasks,
-                          "ncores", ncores,
-                          "nnodes", nnodes,
-                          "walltime", walltime);
-
-    if (msg == NULL) {
-        flux_log_error (h, "failed to create state change event");
-        free (topic);
-        return;
+    if (snprintf (topic, sizeof (topic), "cmb.pub.wreck.state.%s", job->state)
+                                                        >= sizeof (topic)) {
```

Comment:
Thanks, I'll add a comment there.

I didn't see the need for the extra JSON decode/encode.  It could certainly be done that way though!

Reply:
Sorry, I did guess *why* you did it that way, and it seemed smart to me! I was just wondering if this was different enough from other "RPCs" that it might confuse users of the interface. However, if the interface is hidden behind an API call or if this form of RPC topic becomes useful as standard practice, I don't see a problem with it (threw me at first though)

Reply:
I did add a little note about the topic string transformation just now.  Hopefully that helps?

Reply:
Oh sorry, more generally I was thinking that to generalize this we might want to implement some API functions for event publishing.  (Currently we encode an even and pass it to `flux_send()`):
```c
// future is fulfilled once sequence is assigned
flux_future_t *flux_event_publish_seq (flux_t *h, const char *topic, const char *json_str);

// fire and forget
int flux_event_publish (flux_t *h, const char *topic, const char *json_str);
```
Then whatever libertiy we're taking behind the scenes with topic strings is just an implementation detail where a comment will suffice.

Updating the API seems like its own PR though.   Am I still missing the point of your question?

Reply:
Yeah, that satisfies my curiosity.
In this case updating the API is definitely its own PR. However, in future cases where we aren't just updating throwaway code, it is much preferable IMO to add new service, add new API calls, add users of the new API calls, rather than adding API calls after the fact and having to go through and find/update all callers. (i.e. updating the API isn't *always* going to be its own PR)
(After writing that down it seems obvious, sorry!)

Reply:
At the time I was thinking of it as experimental and maybe not useful beyond this module.  But having written down those proposed API calls, maybe I'm convincing myself it's not a bad idea?  Anwyay, if you're OK with deferring to another PR, let's do that.  I'll open an issue to track it.